p2p-tests/contributor-fork-replication-1-test.ts

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env -S node --require ts-node/register/transpile-only --require tsconfig-paths/register
+
+// Copyright © 2021 The Radicle Upstream Contributors
+//
+// This file is part of radicle-upstream, distributed under the GPLv3
+// with Radicle Linking Exception. For full terms see the included
+// LICENSE file.
+
+import * as path from "path";
+import { strict as strictAssert } from "assert";
+
+import * as source from "proxy-client/source";
+import { sleep } from "ui/src/sleep";
+import {
+  RadicleProxy,
+  UpstreamSeed,
+  commit,
+  getLatestCommitSha,
+  pushRad,
+  radCli,
+  runTestcase,
+  withRetry,
+} from "./lib/p2p";
+
+// Test that updates to a project from the contributor are replicated back to
+// the maintainer via a seed.
+//
+// In this testcase we stop and start the peers between issuing new commands.
+async function testcase(dataPath: string) {
+  const project = {
+    name: "my-fancy-project",
+    urn: "rad:git:hnrkrhhs1goaawo7db1gpyct8hd7mif5q8c3o",
+  };
+
+  const seed = new UpstreamSeed({
+    name: "seed",
+    ipAddress: "10.0.0.1",
+    project: project.urn,
+    dataPath,
+  });
+
+  const maintainer = new RadicleProxy({
+    name: "maintainer",
+    ipAddress: "10.0.0.101",
+    seed: seed.seedAddress,
+    dataPath,
+  });
+
+  const contributor = new RadicleProxy({
+    name: "contributor",
+    ipAddress: "10.0.0.102",
+    seed: seed.seedAddress,
+    dataPath,
+  });
+
+  maintainer.start();
+
+  // Maintainer creates a new project.
+  await withRetry(async () => {
+    await maintainer.proxyClient.project.create({
+      repo: {
+        type: "new",
+        path: maintainer.checkoutPath,
+        name: project.name,
+      },
+      description: "",
+      defaultBranch: "main",
+    });
+  });
+
+  await maintainer.stop();
```

Comment:
I don’t think it makes sense to have the maintainer not connected when the project is created. This is tested in other places and requires us to restart the maintainer. Instead, start the maintainer and seed at the beginning and only shutdown the maintainer later.

Reply:
75326e0a34aa781b1faee3416089c00ece9235a1